File: tests/fixtures/vulnerable.kt
// vulnerable.kt — Kotlin fixture with intentional security vulnerabilities.
// Used by kotlin-fixtures.vitest.ts to verify the Kotlin scanner detects all covered types.

package com.example.vulnerable

import java.util.Random
import java.security.MessageDigest
import android.database.sqlite.SQLiteDatabase
import java.io.File

class VulnerableRepository(private val db: SQLiteDatabase) {

    // SECRET_HARDCODED — hardcoded API key in source
    val apiKey = "sk-liveabcdef1234567890secretkey"

    // SECRET_HARDCODED — hardcoded bearer token
    val secretToken = "Bearer eyJhbGciOiJIUzI1NiJ9.payload.sig"

    // INSECURE_RANDOM — java.util.Random used for security-sensitive value
    fun generateSessionId(): Int {
        val rng = Random()
        return rng.nextInt()
    }

    // WEAK_CRYPTO — MD5 MessageDigest
    fun hashPassword(password: String): ByteArray {
        val md = MessageDigest.getInstance("MD5")
        return md.digest(password.toByteArray())
    }

    // WEAK_CRYPTO — SHA-1 MessageDigest
    fun legacyHash(data: String): ByteArray {
        val md = MessageDigest.getInstance("SHA-1")
        return md.digest(data.toByteArray())
    }

    // SQL_INJECTION — rawQuery with string concatenation
    fun getOrdersByUser(userId: String): android.database.Cursor {
        return db.rawQuery("SELECT * FROM orders WHERE user_id = " + userId, null)
    }

    // SQL_INJECTION — rawQuery with string interpolation
    fun getProductById(productId: String): android.database.Cursor {
        return db.rawQuery("SELECT * FROM products WHERE id = ${productId}", null)
    }

    // PATH_TRAVERSAL — File() with user-controlled path argument
    fun readUserFile(path: String): String {
        val file = File(path)
        return file.readText()
    }

    // INSECURE_SHARED_PREFS — storing password in SharedPreferences
    fun cacheCredentials(prefs: android.content.SharedPreferences, pass: String) {
        prefs.edit().putString("password", pass).apply()
    }

    // INSECURE_SHARED_PREFS — storing token in SharedPreferences
    fun cacheToken(prefs: android.content.SharedPreferences, tok: String) {
        prefs.edit().putString("token", tok).apply()
    }

    // PERFORMANCE_N_PLUS_ONE — database query inside a loop (deeply indented)
    fun processOrders(orderIds: List<String>) {
        orderIds.forEach { id ->
                db.rawQuery("SELECT * FROM orders WHERE id = ?", arrayOf(id))
        }
    }
}

File: tests/fixtures/vulnerable.rs
// vulnerable.rs — Rust fixture with intentional security vulnerabilities.
// Used by rust-fixtures.vitest.ts to verify the Rust scanner detects all covered types.

use std::process::Command;
use std::fs;

fn main() {
    // BUFFER_OVERFLOW: raw pointer dereference inside unsafe block (multi-line)
    let value: i32 = 42;
    let ptr: *const i32 = &value;
    unsafe {
        let x = *ptr;
        println!("{}", x);
    }

    // BUFFER_OVERFLOW: mem::transmute usage
    let y: u64 = std::mem::transmute(value);
    println!("{}", y);

    // BUFFER_OVERFLOW: mem::forget
    let s = String::from("hello");
    std::mem::forget(s);

    // SECRET_HARDCODED: hardcoded API key
    let api_key = "sk_live_FAKE_TEST_FIXTURE_NOT_REAL";

    // SECRET_HARDCODED: hardcoded token assignment
    let token = "secret_token_abc123456789xyz";

    // INSECURE_RANDOM: rand::random() for non-crypto use
    let session_id: u64 = rand::random();

    // WEAK_CRYPTO: MD5 usage via md5 crate
    use md5::compute;
    let _digest = md5::compute(b"hello world this is a test");

    // COMMAND_INJECTION: user-controlled variable passed to Command::new
    let user_cmd = "ls";
    let _output = Command::new(user_cmd).output().unwrap();

    // PATH_TRAVERSAL: fs::read_to_string with user-supplied path variable
    let user_path = "/etc/passwd";
    let _contents = fs::read_to_string(user_path).unwrap_or_default();

    // UNSAFE_DESERIALIZATION: serde_json on raw user input
    let body = r#"{"key":"value"}"#;
    let _data: serde_json::Value = serde_json::from_str(body).unwrap();

    println!("{} {} {}", api_key, token, session_id);
}
